pkg/ops/mock_ops.go

```diff
@@ -0,0 +1,433 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ops
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math"
+	"net"
+	"runtime"
+	"syscall"
+
+	"github.com/appc/cni/pkg/ns"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/onsi/ginkgo"
+)
+
+//*****************************************************************************
+
+type testLink struct {
+	link  netlink.Link
+	peer  *testLink
+	addrs []netlink.Addr
+}
+
+type testNamespace struct {
+	ops      *TestNetOps
+	nsid     int
+	nspath   string
+	nextLink int
+	links    map[int]*testLink
+}
+
+type TestNetOps struct {
+	curNamespace *testNamespace
+	namespaces   map[int]*testNamespace
+	nextNsid     int
+}
+
+func newTestNetOps() *TestNetOps {
+	ops := &TestNetOps{
+		namespaces: make(map[int]*testNamespace),
+		nextNsid:   math.MaxInt32,
+	}
+	if _, err := ops.newNamespace(); err != nil {
+		panic("failed to add default namespace")
+	}
+	return ops
+}
+
+func (o *TestNetOps) Log(format string, args ...interface{}) {
+	ginkgo.GinkgoT().Log(fmt.Sprintf(format, args...))
+}
+
+func (o *TestNetOps) getLinkNamespace(link netlink.Link) (*testNamespace, error) {
+	nsid := o.curNamespace.nsid
+	if link != nil {
+		base := link.Attrs()
+		if base.Namespace != nil {
+			switch base.Namespace.(type) {
+			case netlink.NsPid:
+				return nil, syscall.Errno(syscall.EOPNOTSUPP) // unimplemented
+			case netlink.NsFd:
+				nsfd := uint32(base.Namespace.(netlink.NsFd))
+				if nsfd > math.MaxInt32 {
+					return nil, syscall.Errno(syscall.ERANGE)
+				}
+				nsid = int(nsfd)
+			}
+		}
+	}
+	ns, ok := o.namespaces[nsid]
+	if !ok {
+		return nil, syscall.Errno(syscall.ENOENT)
+	}
+	return ns, nil
+}
+
+func (o *TestNetOps) findNamespace(nspath string, nsid int) (*testNamespace, error) {
+	if nsid >= 0 {
+		if ns, ok := o.namespaces[nsid]; ok {
+			return ns, nil
+		}
+	} else {
+		for _, ns := range o.namespaces {
+			if ns.nspath == nspath {
+				return ns, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("Namespace %s / %d not found", nspath, nsid)
+}
+
+func (o *TestNetOps) newNamespace() (*testNamespace, error) {
+	nspath := fmt.Sprintf("/some/fake/path/%d", o.nextNsid)
+	ns := &testNamespace{
+		ops:      o,
+		nsid:     o.nextNsid,
+		nspath:   nspath,
+		nextLink: 1,
+		links:    make(map[int]*testLink),
+	}
+	o.namespaces[o.nextNsid] = ns
+	o.nextNsid--
+	o.curNamespace = ns
+
+	// Add loopback interface
+	o.LinkAdd(&netlink.Device{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: "lo",
+		},
+	})
+
+	return ns, nil
+}
+
+func (o *TestNetOps) getLink(link netlink.Link, name string, index int) (*testLink, *testNamespace, error) {
+	// get link from given namespace; otherwise looks in current namespace
```

Comment:
I suggest to split the fallback case out here so we have `getLink(..)` for the current NS, and `getLinkNs(...)` for a given NS